### ISSUE TITLE
Consul  Catalog Api support to query by tags Or pageing query

### ISSUE THREAD

**Issue by wangyushuai** [hashicorp/consul#9422]
#### Feature Description

When using Catalog Api (**/v1/catalog/services**) to query the service list, I want to filter the list by Tag, because the total number of lists is very large, and I hope to support paging query；

#### Use Case(s)

curl /v1/catalog/services?filter='tag==[app=dashboard]'

**Comment by nugend** [comment]
Would love to have this. Also, just being able to query _all_ service health statuses by tag without first specifying a service id would be really nice too.

**Comment by dekimsey** [comment]
I'm upvoting this because we are running into issues with Prometheus pulling the whole service catalog and trying to set watches on each service so it can perform additional checks.

Right now, much of the data I need to make decisions on which services to monitor is in the `ServiceMeta`. However, the `/v1/catalog/services` API does not expose the filtering capabilities to par down the response. So my applications are having to enumerate each service individually in the catalog, which is expensive.

Since filter appears to have replace the NodeMeta/ServiceMeta parameters in other parts of the API, it'd be nice to have support here as well.

I spent some time digging into Consul's source to see how this might pan out. It's more involved than a change I'm comfortable with proposing at the moment. My first pass involved an update to the api, the rpc, and the store lookup because the current implementation takes the assumption (`{"service1: ["tags", ...] }` response) all the way down to the data-store lookup. I don't know if a change that deep would be accepted by the project, so I kinda stopped there. The end-result of the change would mean the RPC calls are somewhat larger since they'd have the full ServiceNode data response instead of the stripped down tags only response.

Thanks!

Note: There is some related discussion on making the catalog lookups more effective in prometheus, see prometheus/prometheus#7626. I believe adding filtering support on the catalog api would help in this effort.

**Comment by dnephin** [comment]
@dekimsey Thank you for your interest in this issue! I was curious about why this one endpoint did not have filtering (most endpoints do support the `filter` parameter). I see what you are saying about this response data being in a different format. I guess that's why filtering was not added initially.

I did a quick pass to see what might be required to make this work. You can see those changes here: https://github.com/hashicorp/consul/compare/dnephin/catalog-service-list-filter. As it says in the commit message this is missing test changes and docs. I haven't testing it at all, but I would expect something like this to work. 

It did require changing a couple state store functions, but thankfully the only other callers to those functions are a couple tests, so not too big of a change there. I was able to remove what looked like some duplicate logic into the `servicesTagsByName` function. The RPC response should remain the same because we are aggregating and filtering on the server side before returning the response. `servicesTagsByName` is now called from the RPC endpoint after filtering (where as previously it was done in the state store functions).

If you (or anyone!) are still interested in making this change, please do feel free to take the commit I have started and work from that. I'd also be happy to answer any questions you might have to help get it implemented. The commit message has the steps I took and 3 TODOs that should guide next steps.

**Comment by dekimsey** [comment]
@dnephin Nice! Thank you for looking into it. I was a bit concerned about the use of ServiceNode, as I wasn't sure if I was misusing the data model. But learning I wasn't too far off in my own implementation is good stuff :).

I'm going to try to carve out some time this week and see if I can take this the rest of the way.

**Comment by delamart** [comment]
Really looking forward to this feature. It would solve a major requirement for our deployments (being able to find services based on meta data). Hope to see the PR merged soon 🙂

**Comment by dekimsey** [comment]
Sorry @delamart I lost track of this, thank you for the reminder. I'm going to work on cleaning up my PR. I think it was only the tests that needed some tweaks based on feedback.

**Comment by delamart** [comment]
Hi @dekimsey any News on this ? Let me know if I can help in any way.

**Comment by dekimsey** [comment]
@delamart I rebased and addressed the last remaining FIXME I had written. At this point, I'm just waiting on feedback.

**Comment by kisunji** [comment]
This feature has been merged to `main` and should be released in Consul 1.14. Thank you @dekimsey for the PR!
